core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -391,20 +391,23 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     val unauthorizedTopicResponses = mutable.Map[TopicPartition, PartitionResponse]()
     val nonExistingTopicResponses = mutable.Map[TopicPartition, PartitionResponse]()
+    val invalidRequestResponses = mutable.Map[TopicPartition, PartitionResponse]()
     val authorizedRequestInfo = mutable.Map[TopicPartition, MemoryRecords]()
 
     for ((topicPartition, memoryRecords) <- produceRequest.partitionRecordsOrFail.asScala) {
       if (!authorize(request.session, Write, Resource(Topic, topicPartition.topic, LITERAL)))
         unauthorizedTopicResponses += topicPartition -> new PartitionResponse(Errors.TOPIC_AUTHORIZATION_FAILED)
       else if (!metadataCache.contains(topicPartition))
         nonExistingTopicResponses += topicPartition -> new PartitionResponse(Errors.UNKNOWN_TOPIC_OR_PARTITION)
+      else if (request.header.apiVersion() < 7 && memoryRecords.batches().asScala.exists(_.compressionType() == CompressionType.ZSTD))
```

Comment:
Hmm.. The `ProduceRequest` was already constructed, so wouldn't `ProduceRequest.validateRecords` already have failed? I don't see much choice but to move some of the validation logic outside of the constructor. Maybe instead we can do that validation here?

Reply:
> Hmm.. The `ProduceRequest` was already constructed, so wouldn't `ProduceRequest.validateRecords` already have failed?

Well, if the user is using the java client then `ProduceRequest` with api version < 7 compressed with zstd would be already failed. This validation logic defends the other cases - that is, for example, the 3rd party clients without the validation logic.

> I don't see much choice but to move some of the validation logic outside of the constructor. Maybe instead we can do that validation here?

I'm sorry but I can't understand what you mean exactly; The validation condition here is exactly same with that of `ProduceRequest#validateRecords` (api version < 7, compressed with zstd.) Do you mean remove the validation logic in `ProduceRequest#validateRecords` and validate here only?

Reply:
> I'm sorry but I can't understand what you mean exactly; The validation condition here is exactly same with that of ProduceRequest#validateRecords (api version < 7, compressed with zstd.) Do you mean remove the validation logic in ProduceRequest#validateRecords and validate here only?

Yes, sorry if I wasn't clear. As long as the validation is in the `ProduceRequest` constructor, then we will fail before the handler is invoked and we will not be able to return the right error codes. So at least for the `ProduceRequest(Struct)` constructor, I think we need to take out the call to `validateRecords` and move it here instead. Does that make sense?

Reply:
So, you mean...

1. Double validation is overkill.
2. Giving the user accurate error message has high priority.

Therefore, conclusion: **moving the validation logic from `ProducerRequest#validateRecords` into `KafkaApis#handleProduceRequest`  is better.** Since we already have validation logic in `KafkaApis#handleProduceRequest`, removing it from `ProducerRequest#validateRecords` is the right direction.

Do I understand correctly?

Reply:
I think I am not explaining the problem very clearly. When a message is received from the network, the broker must first parse the request. If there are any problems parsing the request, then the broker does not know how to return an error and simply disconnects. That is what is happening here. 

The issue is that the validation logic we have inside `ProduceRequest.validateRecords` is called during the construction of the `ProduceRequest` instance. This logic is invoked during the parsing of the request. So if any of these checks fail, then we do not return any error code at all. We just disconnect.

So to fix this problem and return a helpful error code, we have to let the `ProduceRequest` construction succeed, which means we have to take that validation out of the constructor. Then we can do the validation here inside `KafkaApis.handleProduceRequest`.

Reply:
AHA! Understood. I completely misunderstood `ProducerRequest#validateRecords` is called only in client side!